File: app/src/main/java/com/rachelleignacio/listbucket/domain/interactors/impl/DeleteListItemInteractorImpl.java
package com.rachelleignacio.listbucket.domain.interactors.impl;

import com.rachelleignacio.listbucket.db.DbInteractor;
import com.rachelleignacio.listbucket.domain.executor.Executor;
import com.rachelleignacio.listbucket.domain.executor.MainThread;
import com.rachelleignacio.listbucket.domain.interactors.base.AbstractInteractor;
import com.rachelleignacio.listbucket.domain.interactors.DeleteListItemInteractor;
import com.rachelleignacio.listbucket.domain.models.ListItem;

/**
 * Created by rachelleignacio on 3/10/17.
 */

public class DeleteListItemInteractorImpl extends AbstractInteractor implements DeleteListItemInteractor {
    private DeleteListItemInteractor.Callback callback;
    private DbInteractor database;
    private ListItem itemToDelete;
    private int listAdapterPosition;

    public DeleteListItemInteractorImpl(Executor threadExecutor, MainThread mainThread,
                                        Callback callback, DbInteractor db, ListItem item, int position) {
        super(threadExecutor, mainThread);
        this.callback = callback;
        this.database = db;
        this.itemToDelete = item;
        this.listAdapterPosition = position;
    }

    @Override
    public void run() {
        database.deleteListItem(itemToDelete);
        mainThread.post(new Runnable() {
            @Override
            public void run() {
                callback.onListItemDeleted(listAdapterPosition);
            }
        });
    }
}

File: app/src/main/kotlin/com/rachelleignacio/listbucket/domain/interactors/impl/DeleteListItemInteractorImpl.kt
package com.rachelleignacio.listbucket.domain.interactors.impl

import com.rachelleignacio.listbucket.db.DbInteractor
import com.rachelleignacio.listbucket.domain.executor.MainThread
import com.rachelleignacio.listbucket.domain.executor.ThreadExecutor
import com.rachelleignacio.listbucket.domain.interactors.DeleteListItemInteractor
import com.rachelleignacio.listbucket.domain.interactors.base.AbstractInteractor
import com.rachelleignacio.listbucket.domain.models.ListItem

/**
 * Created by rachelleignacio on 8/9/17.
 */
class DeleteListItemInteractorImpl(threadExecutor: ThreadExecutor,
                                   mainThread: MainThread,
                                   private val callback: DeleteListItemInteractor.Callback,
                                   private val dbInteractor: DbInteractor,
                                   private val itemToDelete: ListItem,
                                   private val adapterPosition: Int
) : AbstractInteractor(threadExecutor, mainThread), DeleteListItemInteractor {

    override fun run() {
        dbInteractor.deleteListItem(itemToDelete)
        mainThread.post(Runnable { callback.onListItemDeleted(adapterPosition) })
    }
}